openvdb_ax/openvdb_ax/codegen/StandardFunctions.cc

```diff
@@ -0,0 +1,1982 @@
+
+        if (elements[0]->getType()->isIntegerTy()) {
+           arithmeticConversion(elements, LLVMType<double>::get(B.getContext()), B);
+        }
+
+        // the following is always done at fp precision
+        llvm::Value* one = llvm::ConstantFP::get(len->getType(), 1.0);
+        llvm::Value* oneDividedByLength = B.CreateFDiv(one, len);
+
+        elements[0] = B.CreateFMul(elements[0], oneDividedByLength);
+        elements[1] = B.CreateFMul(elements[1], oneDividedByLength);
+        elements[2] = B.CreateFMul(elements[2], oneDividedByLength);
+
+        B.CreateStore(elements[0], ptrs[0]);
+        B.CreateStore(elements[1], ptrs[1]);
+        B.CreateStore(elements[2], ptrs[2]);
+
+        return nullptr;
+    };
+
+    static auto norm = [](auto out, auto in) {
+        using VecType = typename std::remove_pointer<decltype(out)>::type;
+        using ElementT = typename openvdb::VecTraits<VecType>::ElementType;
+        *out = *in; // copy
+        out->normalize(ElementT(0.0));
+    };
+
+    using NormalizeD = void(openvdb::math::Vec3<double>*,openvdb::math::Vec3<double>*);
+    using NormalizeF = void(openvdb::math::Vec3<float>*,openvdb::math::Vec3<float>*);
+    using NormalizeI = void(openvdb::math::Vec3<double>*, openvdb::math::Vec3<int32_t>*);
+
+    return FunctionBuilder("normalize")
+        .addSignature<NormalizeD, true>(generate, (NormalizeD*)(norm))
+        .addSignature<NormalizeF, true>(generate, (NormalizeF*)(norm))
+        .addSignature<NormalizeI, true>(generate, (NormalizeI*)(norm))
+        .addDependency("length")
+        .setArgumentNames({"v"})
+        .addParameterAttribute(0, llvm::Attribute::NoAlias)
+        .addParameterAttribute(1, llvm::Attribute::ReadOnly)
+        .addFunctionAttribute(llvm::Attribute::NoUnwind)
+        .addFunctionAttribute(llvm::Attribute::AlwaysInline)
+        .setConstantFold(op.mConstantFoldCBindings)
+        .setPreferredImpl(op.mPrioritiseIR ? FunctionBuilder::IR : FunctionBuilder::C)
+        .setDocumentation("Returns the normalized result of the given vector.")
+        .get();
+}
+
+inline FunctionGroup::Ptr axlerp(const FunctionOptions& op)
+{
+    static auto generate =
+        [](const std::vector<llvm::Value*>& args,
+           llvm::IRBuilder<>& B) -> llvm::Value*
+    {
+        assert(args.size() == 3);
+        llvm::Value* a = args[0];
+        llvm::Value* b = args[1];
+        llvm::Value* x = args[2];
+        llvm::Value* one = llvm::ConstantFP::get(a->getType(), 1.0);
+        llvm::Value* result = binaryOperator(one, x, ast::tokens::MINUS, B);
+        result = binaryOperator(result, a, ast::tokens::MULTIPLY, B);
+        llvm::Value* right = binaryOperator(x, b, ast::tokens::MULTIPLY, B);
+        result = binaryOperator(result, right, ast::tokens::PLUS, B);
+        return result;
+    };
+
+    static auto lerp = [](auto a, auto b, auto x) -> auto {
+        using ValueT = decltype(a);
+        return (ValueT(1.0) - x) * a + x * b;
```

Comment:
```a + (b -a) * x```
is a better formulation, or you have
```
lerp(x, x, t) != x
```
which is surprising.

Reply:
The problem with your version is that `lerp(a,b,1.0f) != b`. However our version is also only monotonic when `a < 0 && b < 0` or `a > 0 and b > 0`... I took a further look at this and turns out the standard introduced lerp in c++20... here's how clang implements it:

``` 
_Fp __lerp(_Fp __a, _Fp __b, _Fp __t) noexcept {
    if ((__a <= 0 && __b >= 0) || (__a >= 0 && __b <= 0))
        return __t * __b + (1 - __t) * __a;

    if (__t == 1) return __b;
    const _Fp __x = __a + __t * (__b - __a);
    if (__t > 1 == __b > __a)
    	return __b < __x ? __x : __b;
    else
    	return __x < __b ? __x : __b;
}
```

https://github.com/llvm/llvm-project/blob/d1dbda10cefeaa124e28eb289cdc92c049c3d973/libcxx/include/cmath#L620

I mean... feels a bit overkill, but this does resolve all cases. And could always then add a `fastlerp` or somesuch.